pkg/controller/atlasprivateendpoint/privateendpoint.go

```diff
@@ -0,0 +1,220 @@
+		inProgress = isInterfaceInProgress(akoPEInterface.Status())
+		var err error
+
+		if !existInAtlas {
+			gcpProjectID := getGCPProjectID(akoPrivateEndpoint, akoPEInterface.InterfaceID())
+			_, err = r.privateEndpointService.CreatePrivateEndpointInterface(
+				ctx.Context,
+				projectID,
+				akoPrivateEndpoint.Spec.Provider,
+				akoPrivateEndpoint.Status.ServiceID,
+				gcpProjectID,
+				akoPEInterface,
+			)
+			if err != nil {
+				return r.terminate(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointReady, workflow.PrivateEndpointFailedToCreate, err)
+			}
+
+			inProgress = true
+		}
+	}
+
+	for _, atlasPEInterface := range atlasPEService.EndpointInterfaces() {
+		akoPEInterfaces := akoPEService.EndpointInterfaces()
+		akoPEInterface := akoPEInterfaces.Get(atlasPEInterface.InterfaceID())
+		wasDeleted := akoPEInterface == nil
+		inProgress = isInterfaceInProgress(atlasPEInterface.Status())
+		var err error
+
+		if wasDeleted && atlasPEInterface.Status() != privateendpoint.StatusDeleting {
+			err = r.privateEndpointService.DeleteEndpointInterface(ctx.Context, projectID, akoPrivateEndpoint.Spec.Provider, akoPrivateEndpoint.Status.ServiceID, atlasPEInterface.InterfaceID())
+			if err != nil {
+				return r.terminate(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointReady, workflow.PrivateEndpointFailedToDelete, err)
+			}
+
+			inProgress = true
+
+			continue
+		}
+
+		if hasInterfaceFailed(atlasPEInterface.Status()) {
+			return r.terminate(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointReady, workflow.PrivateEndpointFailedToConfigure, errors.New(atlasPEInterface.ErrorMessage()))
+		}
+	}
+
+	if !inProgress && reflect.DeepEqual(akoPEService.EndpointInterfaces(), atlasPEService.EndpointInterfaces()) {
+		return r.ready(ctx, akoPrivateEndpoint, atlasPEService)
+	}
+
+	return r.inProgress(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointReady, workflow.PrivateEndpointUpdating, "Private Endpoints are being updated")
+}
+
+func (r *AtlasPrivateEndpointReconciler) getPrivateEndpointService(
+	ctx context.Context,
+	projectID string,
+	akoPrivateEndpoint *akov2.AtlasPrivateEndpoint,
+) (privateendpoint.EndpointService, error) {
+	if akoPrivateEndpoint.Status.ServiceID != "" {
+		return r.privateEndpointService.GetPrivateEndpoint(ctx, projectID, akoPrivateEndpoint.Spec.Provider, akoPrivateEndpoint.Status.ServiceID)
+	}
+
+	endpointServices, err := r.privateEndpointService.ListPrivateEndpoints(ctx, projectID, akoPrivateEndpoint.Spec.Provider)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, endpointService := range endpointServices {
+		if endpointService.Region() == akoPrivateEndpoint.Spec.Region {
+			return endpointService, err
+		}
+	}
+
+	return nil, nil
+}
+
+func (r *AtlasPrivateEndpointReconciler) deletePrivateEndpoint(
+	ctx context.Context,
+	projectID string,
+	atlasPEService privateendpoint.EndpointService,
+) (privateendpoint.EndpointService, error) {
```

Comment:
why does a deletion return an object from a different type? It is confusing

Reply:
I read it wrong, but I still do not understand why we need to pass the `atlasPEService privateendpoint.EndpointService` and return a `privateendpoint.EndpointService`, such service should be part of `r *AtlasPrivateEndpointReconciler` already no?
And even if it were not, why are we retuning it later?

I would expect this signature instead.

```go
func (r *AtlasPrivateEndpointReconciler) deletePrivateEndpoint(ctx context.Context, projectID string) error {
```

But I might be mossing something here.

Reply:
Got you now. The reason here is, one more time, there's a special case added by "yet another" Atlas strange behavior.
1. We have two parts that need to be deleted and both need to wait until finished (status `DELETING`)
2. For AWS, deleting the interface also deletes a service automatically, for the other providers we delete the interface and then delete the service.

Because of these reasons, we return the entity so we can watch its status until the action is completed.

Reply:
ok